### PR TITLE
Fix-1617

### PR DESCRIPTION
**Pull request by loonydev**
Result of grun. Close https://github.com/antlr/grammars-v4/issues/1617

```
grun Scala compilationUnit  examples/class.txt -tree
(compilationUnit package (qualId xxx . yy . zz . aa . bb . commons . transformation) (topStatSeq (topStat (import_ import (importExpr (stableId xxx . yy . zz . aa . bb . commons . constants . Forms)))) (topStat (import_ import (importExpr (stableId xxx . yy . zz . aa . core . configuration . KingTransformation)))) (topStat (import_ import (importExpr (stableId xxx . yy . zz . aa . core . engine . ProcessStatus)))) (topStat (import_ import (importExpr (stableId xxx . yy . zz . aa . core . transformation . TransformationTrait)))) (topStat (import_ import (importExpr (stableId org . apache . spark . sql . DataFrame)))) (topStat (import_ import (importExpr (stableId org . apache . spark . sql . functions) . _))) (topStat (tmplDef class (classDef AddingHoldingFlag1 (classParamClauses (classParamClause ( (classParams (classParam val transformation : (paramType (type_ (infixType (compoundType (annotType (simpleType (stableId KingTransformation))))))))) ))) (classTemplateOpt (templateBody { (templateStat (def def (funDef (funSig runTransformation paramClauses) : (type_ (infixType (compoundType (annotType (simpleType (stableId DataFrame)))))) = (expr (expr1 (postfixExpr (infixExpr (prefixExpr (blockExpr { (block (blockStat (expr1 (postfixExpr (infixExpr (prefixExpr (simpleExpr1 (stableId ???)))))))) }))))))))) })))))))
```
I'm not sure if it's good fix.

### CODE REVIEW

**Comment by loonydev** [comment]
Don't merge yet

**Comment by loonydev** [comment]
Looks good. Used this [link](https://github.com/scala/scala/blob/2.13.x/spec/13-syntax-summary.md) for change grammar.

Yes, I use duplicate of code, but if I use 'or', antlr drop error about recursive. 
Maybe you know better fix for it?

**Comment by loonydev** [comment]
new result of class test (grun).

```
(compilationUnit package (qualId xxx . yy . zz . aa . bb . commons . transformation) (topStatSeq (topStat (import_ import (importExpr (stableId (path xxx . yy . zz . aa . bb . commons . constants) . Forms)))) (topStat (import_ import (importExpr (stableId (path xxx . yy . zz . aa . core . configuration) . KingTransformation)))) (topStat (import_ import (importExpr (stableId (path xxx . yy . zz . aa . core . engine) . ProcessStatus)))) (topStat (import_ import (importExpr (stableId (path xxx . yy . zz . aa . core . transformation) . TransformationTrait)))) (topStat (import_ import (importExpr (stableId (path org . apache . spark) . sql) . (importSelectors { (importSelector DataFrame) , (importSelector Row) , (importSelector functions) })))) (topStat (import_ import (importExpr (stableId (path org . apache . spark . sql) . functions) . _))) (topStat (tmplDef class (classDef AddingHoldingFlag1 (classParamClauses (classParamClause ( (classParams (classParam val transformation : (paramType (type_ (infixType (compoundType (annotType (simpleType (stableId KingTransformation))))))))) ))) (classTemplateOpt extends (classTemplate (classParents (constr (annotType (simpleType (stableId TransformationTrait))) (argumentExprs (blockExpr { (block (blockStat (def def (funDef (funSig runTransformation paramClauses) : (type_ (infixType (compoundType (annotType (simpleType (stableId DataFrame)))))) = (expr (expr1 (postfixExpr (infixExpr (prefixExpr (simpleExpr (blockExpr { (block (blockStat (def (patVarDef val (patDef (pattern2 (pattern3 (simplePattern (stableId previousDF)))) = (expr (expr1 (postfixExpr (infixExpr (prefixExpr (simpleExpr1 (simpleExpr1 (path ProcessStatus . components)) (argumentExprs ( (exprs (expr (expr1 (postfixExpr (infixExpr (prefixExpr (simpleExpr1 (simpleExpr1 (path transformation . previous . get)) (argumentExprs ( (exprs (expr (expr1 (postfixExpr (infixExpr (prefixExpr (simpleExpr1 (literal 0)))))))) ))))))))) )))))))))))) (blockStat (def (patVarDef val (patDef (pattern2 (pattern3 (simplePattern (stableId outPutDataframe)))) = (expr (expr1 (postfixExpr (infixExpr (prefixExpr (simpleExpr1 (simpleExpr1 (path addingHoldingFlag1)) (argumentExprs ( (exprs (expr (expr1 (postfixExpr (infixExpr (prefixExpr (simpleExpr1 (path previousDF)))))))) ))))) outPutDataframe)))))))) })))))))))) (blockStat (def def (funDef (funSig addingHoldingFlag1 (paramClauses (paramClause ( (params (param previousDF : (paramType (type_ (infixType (compoundType (annotType (simpleType (stableId DataFrame))))))))) )))) : (type_ (infixType (compoundType (annotType (simpleType (stableId DataFrame)))))) = (expr (expr1 (postfixExpr (infixExpr (prefixExpr (simpleExpr (blockExpr { (block (blockStat (def (patVarDef val (patDef (pattern2 (pattern3 (simplePattern (stableId addingFlag)))) = (expr (expr1 (postfixExpr (infixExpr (prefixExpr (simpleExpr1 (simpleExpr1 (path previousDF . withColumn)) (argumentExprs ( (exprs (expr (expr1 (postfixExpr (infixExpr (prefixExpr (simpleExpr1 (path Forms . HOLD_1))))))) , (expr (expr1 (postfixExpr (infixExpr (prefixExpr (simpleExpr1 (simpleExpr1 (simpleExpr1 (simpleExpr1 (path when)) (argumentExprs ( (exprs (expr (expr1 (postfixExpr (infixExpr (infixExpr (infixExpr (prefixExpr (simpleExpr1 (simpleExpr1 (simpleExpr1 (simpleExpr1 (path col)) (argumentExprs ( (exprs (expr (expr1 (postfixExpr (infixExpr (prefixExpr (simpleExpr1 (path Forms . capamer)))))))) ))) . contains) (argumentExprs ( (exprs (expr (expr1 (postfixExpr (infixExpr (prefixExpr (simpleExpr1 (literal "WER")))))))) ))))) && (infixExpr (prefixExpr (simpleExpr1 (simpleExpr1 (path col)) (argumentExprs ( (exprs (expr (expr1 (postfixExpr (infixExpr (prefixExpr (simpleExpr1 (path Forms . FLCOUNT)))))))) )))))) === (infixExpr (prefixExpr (simpleExpr1 (literal 1)))))))) , (expr (expr1 (postfixExpr (infixExpr (prefixExpr (simpleExpr1 (simpleExpr1 (path lit)) (argumentExprs ( (exprs (expr (expr1 (postfixExpr (infixExpr (prefixExpr (simpleExpr1 (literal "YES")))))))) ))))))))) ))) . otherwise) (argumentExprs ( (exprs (expr (expr1 (postfixExpr (infixExpr (prefixExpr (simpleExpr1 (simpleExpr1 (path lit)) (argumentExprs ( (exprs (expr (expr1 (postfixExpr (infixExpr (prefixExpr (simpleExpr1 (literal "NO")))))))) ))))))))) ))))))))) ))))) addingFlag)))))))) }))))))))))) })))))))))))
```

File: scala/Scala.g4 [hunk 1]
```diff
@@ -50,9 +50,14 @@ ids
...
+   : Id ('.' Id)*
+   |(Id '.')? 'this'
+   ;
+
 stableId
-   : (Id | (Id '.')? 'this') '.' Id
-   | (Id '.')? 'super' classQualifier? '.' Id
+   : (path '.' Id) | Id
+   | (Id '.') 'super' classQualifier? '.' Id
    ;
```
Comment (KvanTTT) [inline scala/Scala.g4]:
```g4
stableId
   : (path '.')? Id
   | Id '.' 'super' classQualifier? '.' Id
   ;
```

File: scala/Scala.g4 [hunk 2]
```diff
@@ -157,17 +161,25 @@ infixExpr
...
+   ;
+
+simpleExpr
+   : 'new' (classTemplate | templateBody)
+   | blockExpr
    ;
 
+// Dublicate lines to prevent left-recursive code.
+// can't use (simpleExpr|simpleExpr1) '.' Id
 simpleExpr1
```
Comment (KvanTTT) [inline scala/Scala.g4]: Is it possible to use a more clear name?

File: scala/Scala.g4 [hunk 3]
```diff
@@ -371,15 +381,15 @@ import_
...
+   : stableId* ('.' (Id | '_' | importSelectors))?
    ;
 
 importSelectors
    : '{' (importSelector ',')* (importSelector | '_') '}'
    ;
 
 importSelector
-   : Id ('=>' Id | '=>' '_')
+   : Id ('=>' Id | '=>' '_')?
```
Comment (KvanTTT) [inline scala/Scala.g4]:
```g4
     : Id ('=>' (Id | '_'))?
```

File: scala/examples/class.txt [hunk 4]
```diff
@@ -0,0 +1,22 @@
+package  xxx.yy.zz.aa.bb.commons.transformation
```
Comment (KvanTTT) [inline scala/examples/class.txt]: I suggest using SCALA extension like `.scala` instead of default `.txt`.

File: scala/Scala.g4 [hunk 5]
```diff
@@ -157,17 +161,25 @@ infixExpr
...
-   | (Id '.')? 'this'
+   | path
    | '_'
    | '(' exprs? ')'
-   | ('new' (classTemplate | templateBody) | blockExpr) '.' Id
-   | ('new' (classTemplate | templateBody) | blockExpr) typeArgs
+   | simpleExpr  '.' Id
+   | simpleExpr1 '.' Id
+   | simpleExpr  typeArgs
+   | simpleExpr1 typeArgs
```
Comment (KvanTTT) [inline scala/Scala.g4]:
```g4
    | (simpleExpr | simpleExpr1) '.' Id
    | (simpleExpr | simpleExpr1) typeArgs
```

**Comment by KvanTTT** [comment]
> Don't merge yet

In general, it's better to use [WIP] suffix or [draft pull requests](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

> Yes, I use duplicate of code, but if I use 'or', antlr drop error about recursive.

What rule do you mean?

File: scala/Scala.g4 (hunk 5 above)
Comment (loonydev) [reply scala/Scala.g4]: Can't do this. You can try it, but the result is
`error(119): Scala.g4::: The following sets of rules are mutually left-recursive [simpleExpr1]`

That's why I split it to a different line. I think it some problem in generator.

Cause 
`| (simpleExpr | simpleExpr1) '.' Id` 
equal to 
`| simpleExpr '.' Id | simpleExpr1 '.' Id`

File: scala/Scala.g4 (hunk 2 above)
Comment (loonydev) [reply scala/Scala.g4]: Yes, but in spec https://github.com/scala/scala/blob/2.13.x/spec/13-syntax-summary.md used `SimpleExpr` and `SimpleExpr1`, just want to be compliant with spec.

File: scala/Scala.g4 [hunk 6]
```diff
@@ -50,9 +50,14 @@ ids
    : Id (',' Id)*
    ;
 
+path
+   : Id ('.' Id)*
+   |(Id '.')? 'this'
+   ;
```
[the remark this answers is not in the document]
Comment (loonydev) [reply scala/Scala.g4]: I think, it's looks more complex and still two line.

File: scala/Scala.g4 (hunk 2 above)
Comment (KvanTTT) [reply scala/Scala.g4]: Ok then :)

File: scala/Scala.g4 (hunk 5 above)
Comment (KvanTTT) [reply scala/Scala.g4]: Yes, it takes place if left-recursion is involved.

**Comment by KvanTTT** [comment]
Thanks. What about renaming `scala/examples/class.txt` to `scala/examples/class.scala`?

File: scala/Scala.g4 (hunk 5 above)
Comment (loonydev) [reply scala/Scala.g4]: But why, logical it's the same rule. Looks like just a trick to hide from some check in antlr core. Maybe investigate?)

Comment (KvanTTT) [reply scala/Scala.g4]: Already done. There are several issues in ANTLR repository that are similar to our case, for instance, this: https://github.com/antlr/antlr4/issues/564

**Comment by loonydev** [comment]
I done it, not pushed yet, will do tomorrow.

**Comment by loonydev** [comment]
Done. 

So, what about left-recursive error. I can try to fix it, if it's needed.

**Comment by igvls** [comment]
Hi @loonydev ,

I was testing, I found one more issue with the grammar. 
It fails to parse concatenation expressions. Sample code below,
**`
val schema = StructType(Seq(
  StructField(Fields.Alpha + "_trc", StringType),
  StructField(Fields.Beta + "_trc", DecimalType(10,0)),
  StructField(Fields.Gamma, StringType)
))
`**
It would be great, if you help.
Thanks!

**Comment by loonydev** [comment]
Problem here much stranger. Even simple `val a = "test"+"test"` split to a two branch. But I already check the main logic from spec a few time and looks like "normal flow". IMHO, but it must be in one branch with `patDef`
Screen Shot 2020-01-23 at 11 57 30 AM

**Comment by KvanTTT** [comment]
> So, what about left-recursive error. I can try to fix it, if it's needed.

I think we can leave it as-is for now.

**Comment by loonydev** [comment]
Work well with concat, but I broke a "spec"...
Screen Shot 2020-01-23 at 12 16 14 PM

**Pull request merged**